IBM/CAST
# Feature new/csm node find job

# New CSM API 'csm_node_find_job'

## Purpose
This is a new CSM API to help find a job that ran (or is running) on a node during a specific time frame. 

## Approach
The API has been partially implemented, but at this point in time a user can input a `node_name` along with a time range. Then the results will display any job records that ran on that node during that time frame. 

#### Open Questions and Pre-Merge TODOs
- This API has more flags and options for search that has not been implemented yet. But basic functionality addressing the original issue has been completed. I am pushing this code so others on the team can run this in development code to meet the base needs earlier than a full release and also so that the test team can get an early head start developing tests against this new code. 

## How to Test

1. Set up a cluster.
2. Create some allocations.
3. Run this API. while: a.) select node name of a node where this allocation is b.) select a time frame where this job was active. 
4.) see expected matching job record. 
 (advanced tests could include different allocations with different times, some completed and some still running) passing in different time ranges and seeing the jobs appear or disappear from the APIs returnd records accordingly. 

today the working flags are: 
-n : `node_names`
-s : `search_range_begin`
-S : `search_range_end`

you can see more info by running `-h --help` in cmd line interface for this API

## Screenshots


state of database and jobs: 
```


csmdb_vm=# SELECT an.node_name, an.allocation_id, a.primary_job_id, a.user_name, a.num_nodes, a.begin_time, NULL as end_time 
csmdb_vm-# FROM csm_allocation_node AS an
csmdb_vm-# INNER JOIN csm_allocation AS a ON an.allocation_id = a.allocation_id 
csmdb_vm-# UNION  
csmdb_vm-# SELECT anh.node_name, anh.allocation_id, ah.primary_job_id, ah.user_name, ah.num_nodes, ah.begin_time, ah.end_time 
csmdb_vm-# FROM csm_allocation_node_history AS anh
csmdb_vm-# INNER JOIN csm_allocation_history AS ah ON anh.allocation_id = ah.allocation_id 
csmdb_vm-# ORDER BY node_name, allocation_id ASC NULLS LAST;
    node_name    | allocation_id | primary_job_id | user_name | num_nodes |         begin_time         |          end_time          
-----------------+---------------+----------------+-----------+-----------+----------------------------+----------------------------
 c650f03p31-vm03 |             2 |              1 | root      |         3 | 2012-01-10 15:49:29.183549 | 
 c650f03p31-vm04 |             2 |              1 | root      |         3 | 2012-01-10 15:49:29.183549 | 
 c650f03p31-vm05 |             2 |              1 | root      |         3 | 2012-01-10 15:49:29.183549 | 
 c650f03p31-vm05 |             4 |              4 | root      |         3 | 2012-01-10 16:12:12.434493 | 
 c650f03p31-vm08 |             3 |              2 | root      |         3 | 2012-01-10 16:00:09.001173 | 2012-01-16 17:59:43.784577
 c650f03p31-vm08 |             4 |              4 | root      |         3 | 2012-01-10 16:12:12.434493 | 
 c650f03p31-vm10 |             3 |              2 | root      |         3 | 2012-01-10 16:00:09.001173 | 2012-01-16 17:59:43.784577
 c650f03p31-vm10 |             4 |              4 | root      |         3 | 2012-01-10 16:12:12.434493 | 
 c650f03p31-vm12 |             3 |              2 | root      |         3 | 2012-01-10 16:00:09.001173 | 2012-01-16 17:59:43.784577
(9 rows)

```

so in total i have 3 jobs running. some across multiple nodes. 

below i use the api to apply some filters, first raw database matches, then api output. (just to show)

examples of some pre run tests:
```
csmdb_vm=# SELECT an.node_name, an.allocation_id, a.primary_job_id, a.user_name, a.num_nodes, a.begin_time, NULL as end_time 
FROM csm_allocation_node AS an
INNER JOIN csm_allocation AS a ON an.allocation_id = a.allocation_id 
WHERE (
    (
        node_name = 'c650f03p31-vm10' OR 
        node_name = 'c650f03p31-vm04' 
    ) 
    AND 
    (
        (
            begin_time >= '2012-01-10 16:00:00.000000' AND 
            begin_time <= '2012-01-10 17:05:00.000000'
        )
    )
)
UNION  
SELECT anh.node_name, anh.allocation_id, ah.primary_job_id, ah.user_name, ah.num_nodes, ah.begin_time, ah.end_time 
FROM csm_allocation_node_history AS anh
INNER JOIN csm_allocation_history AS ah ON anh.allocation_id = ah.allocation_id 
WHERE (
    (
        node_name = 'c650f03p31-vm10' OR 
        node_name = 'c650f03p31-vm04' 
    ) 
    AND 
    (
        (
            end_time >= '2012-01-10 16:00:00.000000' AND 
            end_time <= '2012-01-10 17:05:00.000000' 
        ) OR 
        (
            begin_time >= '2012-01-10 16:00:00.000000' AND 
            begin_time <= '2012-01-10 17:05:00.000000'
        )
    )
) 
ORDER BY node_name, allocation_id ASC NULLS LAST;
    node_name    | allocation_id | primary_job_id | user_name | num_nodes |         begin_time         |          end_time          
-----------------+---------------+----------------+-----------+-----------+----------------------------+----------------------------
 c650f03p31-vm10 |             3 |              2 | root      |         3 | 2012-01-10 16:00:09.001173 | 2012-01-16 17:59:43.784577
 c650f03p31-vm10 |             4 |              4 | root      |         3 | 2012-01-10 16:12:12.434493 | 
(2 rows)
```

```
[root@c650f03p31-mgt bin]# ./csm_node_find_job -n c650f03p31-vm04,c650f03p31-vm10 -s "2012-01-10 16:00:00.000000" -S "2012-01-10 17:05:00.000000"
---
Total_Records: 2
RECORD_1:
  node_name:      c650f03p31-vm10
  allocation_id:  3
  primary_job_id: 2
  user_name:      root
  num_nodes:      3
  begin_time:     2012-01-10 16:00:09.001173
  end_time:       2012-01-16 17:59:43.784577
RECORD_2:
  node_name:      c650f03p31-vm10
  allocation_id:  4
  primary_job_id: 4
  user_name:      root
  num_nodes:      3
  begin_time:     2012-01-10 16:12:12.434493
  end_time:       
...
```
